### PR TITLE
Fixed some issues of Backprop* functions and added test for BackpropGetBackwardScores

### PR DESCRIPTION
**Pull request by qindazhu**
Will add other tests in the next PR, just make a PR to commit current fixes, especially the NaN issue.

### CODE REVIEW

File: k2/csrc/fsa_utils.cu [hunk 1]
```diff
@@ -1616,27 +1640,24 @@ Array1<FloatType> BackpropGetForwardScores(
...
       Ragged<FloatType> leaving_arc_deriv(leaving_arc_batch.shape);
       FloatType *leaving_arc_deriv_data = leaving_arc_deriv.values.Data();
 
       K2_EVAL(
-          c, num_arcs, lambda_set_arc_deriv_etc, (int32_t arc_idx)->void {
+          c, leaving_arc_batch.NumElements(), lambda_set_arc_deriv_etc,
+          (int32_t arc_idx)->void {
             int32_t fsas_arc_idx012 = leaving_arc_batch_data[arc_idx];
-            Arc arc = arcs[fsas_arc_idx012];
+            const Arc &arc = arcs[fsas_arc_idx012];
```
Comment (csukuangfj) [inline k2/csrc/fsa_utils.cu:1660]: Why to use a reference here? Is `Arc arc` using local memory?

Comment (qindazhu) [reply k2/csrc/fsa_utils.cu:1660]: i think we don't need to copy to local memory here

**Comment by danpovey** [comment]
Cool!  Merge if you think it's ready.

File: k2/csrc/fsa_utils.cu [hunk 2]
```diff
@@ -1237,6 +1237,17 @@ void BackpropGetArcPost(FsaVec &fsas, Ragged<int32_t> &incoming_arcs,
       });
 }
 
+template <>
```
Comment (csukuangfj) [inline k2/csrc/fsa_utils.cu:1240]: This is a specialization, not an instantiation.

`BackpropGetArcPost` is never compiled.

Fixed in #599.

Comment (qindazhu) [reply k2/csrc/fsa_utils.cu:1240]: Yeah, noted I fixed in PR https://github.com/k2-fsa/k2/pull/600, there may be conflicts with you, sorry.

**Pull request merged**